[Test Visibility] Document Python support for Early Flake Detection and Auto Test Retries (#26472)
DataDog/documentation: content/en/tests/flaky_test_management/auto_test_retries.md

> ### Compatibility

Reviewer: In preview there's some blank space between the "compatibility" and "configuration" sections that doesn't look very nice. Can we, perhaps, try moving the two headers inside the tabs

Author: Oh yeah, you're right. That... looks meh. I'll give it a shot.

[jest-image-snapshot][2] is incompatible with `jest.retryTimes` unless `customSnapshotIdentifier` is passed (see [jest-image-snapshot docs][3]) to `toMatchImageSnapshot`. Therefore, auto test retries do not work unless `customSnapshotIdentifier` is used.

> [1]: https://app.datadoghq.com/ci/settings/test-service

Author: ```suggestion
[1]: https://app.datadoghq.com/ci/settings/test-optimization```

> [1]: https://app.datadoghq.com/ci/settings/test-optimization```

Author: ```suggestion
[1]: https://app.datadoghq.com/ci/settings/test-optimization
```

* `DD_CIVISIBILITY_TOTAL_FLAKY_RETRY_COUNT` - a non-negative number to set the maximum total number of failed tests to retry (default: 1000)

> [1]: https://app.datadoghq.com/ci/settings/test-service

Author: ```suggestion
[1]: https://app.datadoghq.com/ci/settings/test-optimization
```

* `DD_CIVISIBILITY_TOTAL_FLAKY_RETRY_COUNT` - a non-negative number to set the maximum total number of failed tests to retry (default: 1000)

> [1]: https://app.datadoghq.com/ci/settings/test-service

Author: ```suggestion
[1]: https://app.datadoghq.com/ci/settings/test-optimization
```

* `DD_CIVISIBILITY_TOTAL_FLAKY_RETRY_COUNT` - a non-negative number to set the maximum total number of failed tests to retry (default: 1000)

> [1]: https://app.datadoghq.com/ci/settings/test-service

Author: ```suggestion
[1]: https://app.datadoghq.com/ci/settings/test-optimization
```

{{% /tab %}}

{{% tab "Python" %}}

> <div class="alert alert-info">Auto Test Retries is available using the beta of the new pytest plugin. Set the <b>DD_PYTEST_USE_NEW_PLUGIN_BETA</b> environment variable to <b>true</b> to enable it.</div>

Reviewer 2: ```suggestion
<div class="alert alert-info">Auto Test Retries is available using the beta of the new pytest plugin. Set the <code>DD_PYTEST_USE_NEW_PLUGIN_BETA</code> environment variable to <code>true</code> to enable it.</div>
```

### Compatibility

`dd-trace-py >= 2.18.0` (`pytest >= 7.2.0`)

### Configuration

After you have set up Test Optimization, you can configure Auto Test Retries from the [Test Service Settings page][1].

{{< img src="continuous_integration/auto_test_retries_test_settings.png" alt="Auto Test Retries in Test Service Settings." style="width:100%" >}}

The default behavior of the feature is to retry any failing test case up to 5 times. Tests that originally fail either the original setup, teardown, or fixtures in Pytest, are not retried.

> This behavior can be fine-tuned with the following environment variables:

Reviewer 2: ```suggestion
You can fine tune this behavior with the following environment variables:
```

> * `DD_CIVISIBILITY_FLAKY_RETRY_ENABLED` - set to `0` or `false` to explicitly disable retries even if the remote setting is enabled (default: true)

Reviewer 2: ```suggestion
* `DD_CIVISIBILITY_FLAKY_RETRY_ENABLED` - set to `0` or `false` to explicitly disable retries even if the remote setting is enabled (default: `true`)
```

### Compatibility

`dd-trace-py >= 2.18.0` (`pytest >= 7.2.0`)

### Configuration

After you have set up Test Optimization, you can configure Auto Test Retries from the [Test Service Settings page][1].

{{< img src="continuous_integration/auto_test_retries_test_settings.png" alt="Auto Test Retries in Test Service Settings." style="width:100%" >}}

The default behavior of the feature is to retry any failing test case up to five times. Tests that originally fail either the original setup, teardown, or fixtures in Pytest, are not retried.

You can fine tune this behavior with the following environment variables:

* `DD_CIVISIBILITY_FLAKY_RETRY_ENABLED` - set to `0` or `false` to explicitly disable retries even if the remote setting is enabled (default: `true`)
* `DD_CIVISIBILITY_FLAKY_RETRY_COUNT` - a non-negative number to change the maximum number of retries per test case (default: `5`).
* `DD_CIVISIBILITY_TOTAL_FLAKY_RETRY_COUNT` - a non-negative number to set the maximum total number of failed tests to retry (default: `1000`)

> [1]: https://app.datadoghq.com/ci/settings/test-service

Author: ```suggestion
[1]: https://app.datadoghq.com/ci/settings/test-optimization
```

[1]: /tests/setup/
[2]: /tests/explorer/
[3]: /tests/flaky_test_management/early_flake_detection

> [4]: https://app.datadoghq.com/ci/settings/test-service

Author: ```suggestion
[4]: https://app.datadoghq.com/ci/settings/test-optimization
```
